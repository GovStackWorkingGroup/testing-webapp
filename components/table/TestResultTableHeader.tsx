import React, { useCallback, useEffect, useState } from 'react';
import { useIntl } from 'react-intl';
import { RiArrowDownSFill, RiArrowUpSFill } from 'react-icons/ri';
import classNames from 'classnames';
import { ResultTableColumnSortType, ResultTableSortByType, SortResultTableFieldType } from './types';

type Props = {
  handleSorting: (tableSortProperties: ResultTableSortByType) => void;
};

const TestResultTableHeader = ({ handleSorting }: Props) => {
  const [statusSort, setStatusSort] = useState<ResultTableColumnSortType>({ field: 'status',order: null });
  const [uriSort, setUriSort] = useState<ResultTableColumnSortType>({ field: 'uri', order: null });
  const [nameSort, setNameSort] = useState<ResultTableColumnSortType>({ field: 'name', order: null });

  const { formatMessage } = useIntl();
  const format = useCallback(
    (id: string) => formatMessage({ id }),
    [formatMessage]
  );

  useEffect(() => {
    handleSorting({ status: statusSort, uri: uriSort, name: nameSort });
  }, [statusSort, uriSort, nameSort]);

  const handleSortingChange = (field: SortResultTableFieldType) => {
    if (field === 'status') {
      setStatusSort({
        field: 'status',
        order: statusSort.order === 'asc' ? 'desc' : (statusSort.order === 'desc' ? null : 'asc'),
      });
    } else if (field === 'uri') {
      setUriSort({
        field: 'uri',
        order: uriSort.order === 'asc' ? 'desc' : (uriSort.order === 'desc' ? null : 'asc'),
      });
    } else if (field === 'name') {
      setNameSort({
        field: 'name',
        order: nameSort.order === 'asc' ? 'desc' : (nameSort.order === 'desc' ? null : 'asc'),
      });
    }
  };

  return (
    <div className='test-table-header'>
      <div onClick={() => handleSortingChange('status')} data-testid='header-status-label'>
        <p>{format('test_table.status.label')}</p>
        <div className='sort-icon'>
          <RiArrowUpSFill
            className={classNames('sort-icon-up', {
              'sort-icon-active': statusSort.order === 'asc',
            })}
          />
          <RiArrowDownSFill
            className={classNames('sort-icon-down', {
              'sort-icon-active': statusSort.order === 'desc',
            })}
          />
        </div>
      </div>
      <div onClick={() => handleSortingChange('uri')} data-testid='header-category-label'>
        <p>{format('test_table.category.label')}</p>
        <div className='sort-icon'>
          <RiArrowUpSFill
            className={classNames('sort-icon-up', {
              'sort-icon-active': uriSort.order === 'asc',
            })}
          />
          <RiArrowDownSFill
            className={classNames('sort-icon-down', {
              'sort-icon-active': uriSort.order === 'desc',
            })}
          />
        </div>
      </div>
      <div onClick={() => handleSortingChange('name')} data-testid='header-name-label'>
        <p>{format('test_table.name.label')}</p>
        <div className='sort-icon'>
          <RiArrowUpSFill
            className={classNames('sort-icon-up', {
              'sort-icon-active': nameSort.order === 'asc',
            })}
          />
          <RiArrowDownSFill
            className={classNames('sort-icon-down', {
              'sort-icon-active': nameSort.order === 'desc',
            })}
          />
        </div>
      </div>
    </div>
  );
};

export default TestResultTableHeader;
