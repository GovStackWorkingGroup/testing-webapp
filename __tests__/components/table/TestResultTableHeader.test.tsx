import { fireEvent } from '@testing-library/react';
import TestResultTableHeader from '../../../components/table/TestResultTableHeader';
import { render } from '../../test-utils/test-utils';

describe('Unit tests for TestResultTableHeader component:', () => {
  const STATUS_LABEL_TEST_ID = 'header-status-label';
  const CATEGORY_LABEL_TEST_ID = 'header-category-label';
  const NAME_LABEL_TEST_ID = 'header-name-label';

  const mockedHandleSorting = jest.fn();

  it('should render the TestResultTableHeader component and match snapshot', () => {
    const { container } = render(<TestResultTableHeader handleSorting={mockedHandleSorting}/>);

    expect(container).toMatchSnapshot();
  });

  it('should have proper labels', () => {
    const { getByTestId } = render(<TestResultTableHeader handleSorting={mockedHandleSorting}/>);

    expect(getByTestId(STATUS_LABEL_TEST_ID)).toHaveTextContent('Status');
    expect(getByTestId(CATEGORY_LABEL_TEST_ID)).toHaveTextContent('Category');
    expect(getByTestId(NAME_LABEL_TEST_ID)).toHaveTextContent('Name');
  });

  it('should call the sorting function after clicks on the Status header or sort icon', () => {
    const mockedHandleSorting = jest.fn();
    const { getByTestId } = render(<TestResultTableHeader handleSorting={mockedHandleSorting} />);

    expect(mockedHandleSorting).toHaveBeenCalledTimes(1);

    fireEvent.click(getByTestId(STATUS_LABEL_TEST_ID));
    expect(mockedHandleSorting).toHaveBeenCalledTimes(2);
    expect(mockedHandleSorting).toHaveBeenCalledWith({
      'status': { 'field': 'status', 'order': 'asc' },
      'uri': { 'field': 'uri', 'order': null },
      'name': { 'field': 'name', 'order': null }
    });

    fireEvent.click(getByTestId(STATUS_LABEL_TEST_ID));
    expect(mockedHandleSorting).toHaveBeenCalledTimes(3);
    expect(mockedHandleSorting).toHaveBeenCalledWith({
      'status': { 'field': 'status', 'order': 'desc' },
      'uri': { 'field': 'uri', 'order': null },
      'name': { 'field': 'name', 'order': null }
    });

    fireEvent.click(getByTestId(STATUS_LABEL_TEST_ID));
    expect(mockedHandleSorting).toHaveBeenCalledTimes(4);
    expect(mockedHandleSorting).toHaveBeenCalledWith({
      'status': { 'field': 'status', 'order': null },
      'uri': { 'field': 'uri', 'order': null },
      'name': { 'field': 'name', 'order': null }
    });
  });

  it('should call the sorting function after clicks on the Category header or sort icon', () => {
    const mockedHandleSorting = jest.fn();
    const { getByTestId } = render(<TestResultTableHeader handleSorting={mockedHandleSorting} />);

    expect(mockedHandleSorting).toHaveBeenCalledTimes(1);

    fireEvent.click(getByTestId(CATEGORY_LABEL_TEST_ID));
    expect(mockedHandleSorting).toHaveBeenCalledTimes(2);
    expect(mockedHandleSorting).toHaveBeenCalledWith(
      {
        'status': { 'field': 'status', 'order': null },
        'uri': { 'field': 'uri', 'order': 'asc' },
        'name': { 'field': 'name', 'order': null }
      });

    fireEvent.click(getByTestId(CATEGORY_LABEL_TEST_ID));
    expect(mockedHandleSorting).toHaveBeenCalledTimes(3);
    expect(mockedHandleSorting).toHaveBeenCalledWith({
      'status': { 'field': 'status', 'order': null },
      'uri': { 'field': 'uri', 'order': 'desc' },
      'name': { 'field': 'name', 'order': null }
    });

    fireEvent.click(getByTestId(STATUS_LABEL_TEST_ID));
    expect(mockedHandleSorting).toHaveBeenCalledTimes(4);
    expect(mockedHandleSorting).toHaveBeenCalledWith({
      'status': { 'field': 'status', 'order': null },
      'uri': { 'field': 'uri', 'order': null },
      'name': { 'field': 'name', 'order': null }
    });
  });

  it('should call the sorting function after clicks on the Name header or sort icon', () => {
    const mockedHandleSorting = jest.fn();
    const { getByTestId } = render(<TestResultTableHeader handleSorting={mockedHandleSorting} />);

    expect(mockedHandleSorting).toHaveBeenCalledTimes(1);

    fireEvent.click(getByTestId(NAME_LABEL_TEST_ID));
    expect(mockedHandleSorting).toHaveBeenCalledTimes(2);
    expect(mockedHandleSorting).toHaveBeenCalledWith({
      'status': { 'field': 'status', 'order': null },
      'uri': { 'field': 'uri', 'order': null },
      'name': { 'field': 'name', 'order': 'asc' }
    });

    fireEvent.click(getByTestId(NAME_LABEL_TEST_ID));
    expect(mockedHandleSorting).toHaveBeenCalledTimes(3);
    expect(mockedHandleSorting).toHaveBeenCalledWith({
      'status': { 'field': 'status', 'order': null },
      'uri': { 'field': 'uri', 'order': null },
      'name': { 'field': 'name', 'order': 'desc' }
    });

    fireEvent.click(getByTestId(STATUS_LABEL_TEST_ID));
    expect(mockedHandleSorting).toHaveBeenCalledTimes(4);
    expect(mockedHandleSorting).toHaveBeenCalledWith({
      'status': { 'field': 'status', 'order': null },
      'uri': { 'field': 'uri', 'order': null },
      'name': { 'field': 'name', 'order': null }
    });
  });
});
