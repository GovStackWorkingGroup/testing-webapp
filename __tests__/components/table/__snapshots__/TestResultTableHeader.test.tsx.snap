// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Unit tests for TestResultTableHeader component: should render the TestResultTableHeader component and match snapshot 1`] = `
<div>
  <div
    class="test-table-header"
  >
    <div
      data-testid="header-status-label"
    >
      <p>
        Status
      </p>
      <div
        class="sort-icon"
      >
        <svg
          class="sort-icon-up"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <g>
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
            <path
              d="M12 8l6 6H6z"
            />
          </g>
        </svg>
        <svg
          class="sort-icon-down"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <g>
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
            <path
              d="M12 16l-6-6h12z"
            />
          </g>
        </svg>
      </div>
    </div>
    <div
      data-testid="header-category-label"
    >
      <p>
        Category
      </p>
      <div
        class="sort-icon"
      >
        <svg
          class="sort-icon-up"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <g>
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
            <path
              d="M12 8l6 6H6z"
            />
          </g>
        </svg>
        <svg
          class="sort-icon-down"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <g>
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
            <path
              d="M12 16l-6-6h12z"
            />
          </g>
        </svg>
      </div>
    </div>
    <div
      data-testid="header-name-label"
    >
      <p>
        Name
      </p>
      <div
        class="sort-icon"
      >
        <svg
          class="sort-icon-up"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <g>
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
            <path
              d="M12 8l6 6H6z"
            />
          </g>
        </svg>
        <svg
          class="sort-icon-down"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <g>
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
            <path
              d="M12 16l-6-6h12z"
            />
          </g>
        </svg>
      </div>
    </div>
  </div>
</div>
`;
